### PR TITLE
Add `[tool.maturin.include]` and `[tool.maturin.exclude]`

### PR DESCRIPTION
**Pull request by mbrobbel**
Based on the discussion in #1253, this is another attempt to allow users to override ignore files in Python modules by adding `[tool.maturin.include]` and `[tool.maturin.exclude]` similar to [Poetry](https://python-poetry.org/docs/pyproject/#include-and-exclude). This deprecates `[tool.maturin.sdist-include]`.

After some feedback on this approach, I'll add:
- [x] Documentation updates
- [x] Integration tests

Closes #1253

### CODE REVIEW

File: src/pyproject_toml.rs [hunk 1]
```diff
@@ -98,10 +168,24 @@ impl PyProjectToml {
     }
 
     /// Returns the value of `[tool.maturin.sdist-include]` in pyproject.toml
+    #[deprecated(
+        since = "0.14.0",
+        note = "please use `PyProjectToml::include` (<https://github.com/PyO3/maturin/pulls/1255>)"
+    )]
```
Comment (messense) [inline src/pyproject_toml.rs:174]: maturin is mostly used as a cli, so this `deprecated` annotation won't be noticed by users.

File: src/source_distribution.rs [hunk 2]
```diff
@@ -551,20 +553,36 @@ pub fn source_distribution(
...
+            } else {
+                writer.add_file(target, source)?;
+            }
+        }
+        Ok(())
+    };
+
+    #[allow(deprecated)]
     if let Some(include_targets) = pyproject.sdist_include() {
+        warn!("`[tool.maturin.sdist-include]` is deprecated, please use `[tool.maturin.include]`");
```
Comment (messense) [inline src/source_distribution.rs:574]: `warn!` requires settings `RUST_LOG` and will add timestamp to the message, just use `eprintln!` instead.

**Review by messense — COMMENTED** [review]
Thanks a lot for looking into this!

File: src/source_distribution.rs [hunk 3]
```diff
@@ -551,20 +553,38 @@ pub fn source_distribution(
...
+                writer.add_file(target, source)?;
+            }
+        }
+        Ok(())
+    };
+
+    #[allow(deprecated)]
     if let Some(include_targets) = pyproject.sdist_include() {
+        eprintln!(
+            "`[tool.maturin.sdist-include]` is deprecated, please use `[tool.maturin.include]`"
```
Comment (messense) [inline src/source_distribution.rs:575]: We usually also add an emoji

```suggestion
            "⚠️  Warning: `[tool.maturin.sdist-include]` is deprecated, please use `[tool.maturin.include]`"
```

**Review by messense — COMMENTED** [review]
This already looks pretty good, should be ready soon when you are done with tests and documentation changes.

File: tests/run.rs [hunk 4]
```diff
@@ -463,6 +493,33 @@ fn pyo3_mixed_src_layout_sdist() {
     ))
 }
 
+#[test]
+fn pyo3_mixed_include_exclude_src_layout_sdist() {
```
Comment (messense) [inline tests/run.rs:497]:
```suggestion
fn pyo3_mixed_include_exclude_sdist() {
```

File: tests/run.rs [hunk 5]
```diff
@@ -463,6 +493,33 @@ fn pyo3_mixed_src_layout_sdist() {
...
+            "pyo3_mixed_include_exclude-2.1.3/pyo3_mixed_include_exclude/__init__.py",
+            "pyo3_mixed_include_exclude-2.1.3/pyo3_mixed_include_exclude/python_module/__init__.py",
+            "pyo3_mixed_include_exclude-2.1.3/pyo3_mixed_include_exclude/python_module/double.py",
+            "pyo3_mixed_include_exclude-2.1.3/pyproject.toml",
+            "pyo3_mixed_include_exclude-2.1.3/src/lib.rs",
+            // "pyo3_mixed_include_exclude-2.1.3/tests/test_pyo3_mixed_include_exclude.py", excluded
+            "pyo3_mixed_include_exclude-2.1.3/tox.ini",
+        ],
+        None,
+        "sdist-pyo3-mixed-include-exclude-src-layout",
```
Comment (messense) [inline tests/run.rs:519]:
```suggestion
        "sdist-pyo3-mixed-include-exclude",
```

File: test-crates/pyo3-mixed-include-exclude/pyo3-config.txt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+implementation=CPython
```
Comment (messense) [inline test-crates/pyo3-mixed-include-exclude/pyo3-config.txt:1]: You can remove this file since we are not testing the related functionality.

**Review by messense — COMMENTED** [review]
I think we are missing a file list check for wheel, unfortunately we don't have a existing infrastructure for this kind of test yet.

File: src/lib.rs [hunk 7]
```diff
@@ -35,7 +35,7 @@ pub use crate::module_writer::{
     write_dist_info, ModuleWriter, PathWriter, SDistWriter, WheelWriter,
 };
 pub use crate::new_project::{init_project, new_project, GenerateProjectOptions};
-pub use crate::pyproject_toml::PyProjectToml;
+pub use crate::pyproject_toml::{Format, Formats, GlobPattern, PyProjectToml};
```
Comment (messense) [inline src/lib.rs:38]: Perhaps we should make `pyproject_toml` mod `pub`, export `Format` and stuffs in top level doesn't make much sense.

File: tests/common/other.rs [hunk 8]
```diff
@@ -169,6 +171,48 @@ pub fn test_source_distribution(
...
+    let manifest_path = package.as_ref().join("Cargo.toml");
+    let wheel_directory = Path::new("test-crates").join("wheels").join(unique_name);
+
+    let build_options = BuildOptions {
+        out: Some(wheel_directory),
+        cargo: CargoOptions {
+            manifest_path: Some(manifest_path),
+            quiet: true,
+            target_dir: Some(PathBuf::from(
+                "test-crates/targets/test_workspace_cargo_lock",
```
Comment (messense) [inline tests/common/other.rs:188]:
```suggestion
                format!("test-crates/targets/{}", unique_name),
```

File: tests/run.rs [hunk 9]
```diff
@@ -44,6 +44,16 @@ fn develop_pyo3_mixed() {
...
+#[test]
+fn develop_pyo3_mixed_include_exclude() {
+    handle_result(develop::test_develop(
+        "test-crates/pyo3-mixed-include-exclude",
+        None,
+        "develop-pyo3-mixed-include-exclude",
+        false,
+    ));
+}
+
```
Comment (messense) [inline tests/run.rs:56]:
```suggestion
```

File: tests/run.rs [hunk 10]
```diff
@@ -132,6 +142,15 @@ fn editable_pyo3_mixed() {
...
 
+#[test]
+fn editable_pyo3_mixed_include_exclude() {
+    handle_result(editable::test_editable(
+        "test-crates/pyo3-mixed-include-exclude",
+        None,
+        "editable-pyo3-mixed-include-exclude",
+    ));
+}
+
```
Comment (messense) [inline tests/run.rs:153]:
```suggestion
```

File: tests/run.rs [hunk 11]
```diff
@@ -193,6 +212,17 @@ fn integration_pyo3_mixed() {
...
+fn integration_pyo3_mixed_include_exclude() {
+    handle_result(integration::test_integration(
+        "test-crates/pyo3-mixed-include-exclude",
+        None,
+        "integration-pyo3-mixed-include-exclude",
+        false,
+        None,
+    ));
+}
+
```
Comment (messense) [inline tests/run.rs:225]:
```suggestion
```

I think we can just drop `develop`, `editable` and `integration` tests for this new test crate, we only want to test `include` and `exclude`.

File: test-crates/pyo3-mixed-include-exclude/pyproject.toml [hunk 12]
```diff
@@ -0,0 +1,25 @@
+[build-system]
+requires = ["maturin>=0.13,<0.14"]
+build-backend = "maturin"
+
+[project]
+name = "pyo3-mixed-include-exclude"
+classifiers = ["Programming Language :: Python", "Programming Language :: Rust"]
+requires-python = ">=3.7"
+dependencies = ["boltons"]
```
Comment (messense) [inline test-crates/pyo3-mixed-include-exclude/pyproject.toml:9]:
```suggestion
```

File: test-crates/pyo3-mixed-include-exclude/check_installed/check_installed.py [hunk 13]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env python3
+
+from boltons.strutils import slugify
```
Comment (messense) [inline test-crates/pyo3-mixed-include-exclude/check_installed/check_installed.py:3]:
```suggestion
```

File: test-crates/pyo3-mixed-include-exclude/check_installed/check_installed.py [hunk 14]
```diff
@@ -0,0 +1,9 @@
...
+import pyo3_mixed_include_exclude
+
+assert pyo3_mixed_include_exclude.get_42() == 42
+assert slugify("First post! Hi!!!!~1    ") == "first_post_hi_1"
```
Comment (messense) [inline test-crates/pyo3-mixed-include-exclude/check_installed/check_installed.py:7]:
```suggestion
```

**Review by messense — COMMENTED** [review]
We can simplify this test crate.

**Review by messense — APPROVED** [review]
Looks good to me, thanks!

**Comment by messense** [comment]
There are some test failures: https://github.com/PyO3/maturin/actions/runs/3427030079/jobs/5709531358

File: tests/common/other.rs [hunk 15]
```diff
@@ -169,6 +171,49 @@ pub fn test_source_distribution(
...
+        cargo: CargoOptions {
+            manifest_path: Some(manifest_path),
+            quiet: true,
+            target_dir: Some(PathBuf::from(format!(
+                "test-crates/targets/{}",
+                unique_name
+            ))),
+            ..Default::default()
+        },
+        ..Default::default()
```
Comment (messense) [inline tests/common/other.rs:193]: Set `platform_tag` to `Linux` to avoid requiring `patchelf` on Alpine Linux.

**Comment by mbrobbel** [comment]
> Still need to fix the sdist test: https://github.com/PyO3/maturin/actions/runs/3427943992/jobs/5711566272, sdist shouldn't include `.pyc` or `.so` files.

I'm confused about how these files end up in the source distribution and that this fails for the added test but not for the one this was based on? Shouldn't the root `.gitignore` prevent those files from being added?

**Comment by messense** [comment]
I'm not sure, perhaps you can try add the ignores to https://github.com/PyO3/maturin/pull/1255/files#diff-130021135ea4ccb1cd3ba6e252477d65142317ffba1e4a10ad7af791d30c5b12

If you want to debug it, use the `alpine:edge` docker image.

**Comment by messense** [comment]
You can also try modify `.github/workflows/test.yml` and add

```yaml
    - name: Setup tmate session
      uses: mxschmitt/action-tmate@v3
```

to `test-alpine` job and get a ssh shell.

**Comment by messense** [comment]
`cargo package --list --allow-dirty` didn't filter them

```
[
    "Cargo.toml",
    "Cargo.toml.orig",
    "README.md",
    "check_installed/check_installed.py",
    "pyo3_mixed_include_exclude/__init__.py",
    "pyo3_mixed_include_exclude/__pycache__/__init__.cpython-310.pyc",
    "pyo3_mixed_include_exclude/exclude_this_file",
    "pyo3_mixed_include_exclude/include_this_file",
    "pyo3_mixed_include_exclude/pyo3_mixed_include_exclude.cpython-310-x86_64-linux-gnu.so",
    "pyo3_mixed_include_exclude/python_module/__init__.py",
    "pyo3_mixed_include_exclude/python_module/__pycache__/__init__.cpython-310.pyc",
    "pyo3_mixed_include_exclude/python_module/__pycache__/double.cpython-310.pyc",
    "pyo3_mixed_include_exclude/python_module/double.py",
    "pyproject.toml",
    "src/lib.rs",
    "tests/test_pyo3_mixed_include_exclude.py",
    "tox.ini",
]
```

**Comment by mbrobbel** [comment]
Thank you for the help and quick feedback to get this in so quickly!

**Pull request merged**